Add a new scale

AryanGodara opened the pull request:
Resolves #3 (May add more scales if this one works properly)
Review needed, not ready yet

Engil commented on cardio_crumble.opam:
```diff
@@ -13,8 +13,6 @@ depends: [
   "ocaml"
   "dune" {>= "3.3"}
   "odoc" {with-doc}
-  "portmidi"
```
I don't think this should go away?

Engil replied:
(I did not review the rest yet, will take a look maybe later today or this week end! But this is great, thank you very much!)

AryanGodara replied:
Thanks for the review @Engil !
It seems like I deleted this accidently xD while going through all the files, and checking dependencies etc.

AryanGodara replied:
@Engil , just checked, this is weird
When I run,
```
dune build
```
It removes `portmidi` automatically. Is this not being used in the project anymore?

pitag-ha replied:
> The dependencies should be added to dune-project, not the .opam file. Dune seems to be handling dependencies by itself

We've specified that we want Dune to do so by having `(generate_opam_files true)` in the `dune-project`.

@AryanGodara, do you want to fix this for us by adding the two dependencies to the `dune-project` file (and re-adding them to the `.opam`-file)? You could have a look at the [`dune-project` file of the portmidi library](https://github.com/mbacarella/portmidi/blob/main/dune-project#L22-L33) to see the syntax. No pressure! I'm happy to do it myself if you prefer to focus on the scale task.

AryanGodara replied:
No I'd definitely like to work on this issue @pitag-ha !!
Should I open a separate issue for this? Or increase the scope of this current PR?

pitag-ha replied:
> No I'd definitely like to work on this issue @pitag-ha !!

Nice!

> Should I open a separate issue for this? Or increase the scope of this current PR?

I leave that up to you. The cleaner version would be to open a separate PR, merge it, and then rebase this branch over the branch of the other PR. However, as you've said, this change is minimal, so mixing the things here is ok as well. If you think otherwise about git history pedantry, @Engil, let us know!

pitag-ha commented on .ocamlformat:
```diff
@@ -1,3 +1,3 @@
-version=0.24.1
+version=0.25.1
```
Just for some context: the `ocamlformat.0.25` version was released a couple of weeks ago, and we haven't upgraded it yet. You've probably seen an error when running `dune @build` pointing out that you have `ocamlformat.0.25.1` installed, but `ocamlformat.0.24.1` specified in the `.ocamlformat` file. Your solution was to update the `.ocamlformat`-file to make sure the two coincide. Is that right?

If so, the cleaner solution would have been to downgrade your installed `ocamlformat` version to `0.24.1`. That way, we would keep a clean git history: one commit/PR adding the two new scales; and, possibly, later another commit/PR upgrading `ocamlformat` and re-formatting the code base.

All this explanation is just to give you some context. In this project, we don't need to be too pedantic about the git history and can keep your approach mixing the two things in one. @Engil , let me know if you think otherwise!

AryanGodara replied:
Thanks for explaining the issue @pitag-ha !
I have downgraded back to 0.24.1, and will open separate PR for upgrading ocamlformat in the repo!

AryanGodara replied:
And yes, my plan was to concide the ocamlformat versions :)

pitag-ha commented on bin/midi.ml:
```diff
@@ -81,6 +81,34 @@ module Scale = struct
+        failwith
+          "Why on earth is something mod 7 not element of {0,1,2,3,4,5,6}?"
+
+  let pentatonic base_note i =
+    let octave, scale_func = partition i in
+    match scale_func with
+    | 0 -> Char.chr @@ (base_note + (12 * octave))
+    | 1 -> Char.chr @@ (base_note + 2 + (12 * octave))
+    | 2 -> Char.chr @@ (base_note + 4 + (12 * octave))
+    | 3 -> Char.chr @@ (base_note + 7 + (12 * octave))
+    | 4 -> Char.chr @@ (base_note + 9 + (12 * octave))
+    | 5 -> Char.chr @@ (base_note + 12 + (12 * octave))
+    | 6 -> Char.chr @@ (base_note + 14 + (12 * octave))
+    | _ ->
+        failwith
+          "Why on earth is something mod 7 not element of {0,1,2,3,4,5,6}?"
```
Very nice idea to add a pentatonic scale as well!

pitag-ha commented on bin/midi.ml:
```diff
@@ -81,6 +81,34 @@ module Scale = struct
+    | 2 -> Char.chr @@ (base_note + 4 + (12 * octave))
+    | 3 -> Char.chr @@ (base_note + 7 + (12 * octave))
+    | 4 -> Char.chr @@ (base_note + 9 + (12 * octave))
+    | 5 -> Char.chr @@ (base_note + 12 + (12 * octave))
+    | 6 -> Char.chr @@ (base_note + 14 + (12 * octave))
```
Very nice solution to the problem that this new scale has fewer notes than the others!

AryanGodara replied:
Yes, I was going to ask you whether this was okay, I just looped around back to initial note, just higher octave.

pitag-ha commented on .gitignore:
```diff
@@ -3,3 +3,4 @@ _build
 *.merlin
 _opam
 *.events
+.vscode
```
Do you have a global gitignore file? Or could set one up? Given that `.vscode` is user specific rather project specific, it would be more hygienic to add it to your global gitignore. FWIW (for what it's worth), I myself have `.vscode/*` in my global gitignore as well.

AryanGodara replied:
I wasn't doing this, thanks for this tip, it'll save me time in a lot of cloned repos!!

pitag-ha reviewed (commented):
Thanks, @AryanGodara, this is a great first contribution!

Apart from the few comments I have below (and the one about the dependencies is clearly on us), it looks all good to me to be merged.

pitag-ha reviewed (commented):
Thanks for the changes, @AryanGodara!

AryanGodara commented on bin/midi.mli:
```diff
@@ -2,23 +2,22 @@ module Event = Portmidi.Portmidi_event
-
-val message_on :
-  note:char -> timestamp:int32 -> ?volume:char -> unit -> Event.t
+module Device : sig
+  type t
+
+  val create : int -> t
+  val shutdown : t -> (unit, Portmidi.Portmidi_error.t) result
+end
+
+val message_on : note:char -> timestamp:int32 -> ?volume:char -> unit -> Event.t
```
No, I did this manually, to pass the `dune build @fmt` test.
(OCaml-mbr had this test, so I thought of testing for this commit also)

pitag-ha replied:
What does `dune build @fmt` currently return for you? And which version of `ocamlformat` do you have installed now (`$ ocamlformat --version`)?

AryanGodara commented:
@pitag-ha , I've updated the new changes, sorry for the delay!

pitag-ha commented:
> I've updated the new changes,

Thanks! To be a bit clearer on the dependencies: Given that you're fixing the dependencies in a different PR, this PR should not do any modifications on the `.opam` file and the `dune-project` file. Is it clear why?

AryanGodara commented:
> > I've updated the new changes,
> 
> Thanks! To be a bit clearer on the dependencies: Given that you're fixing the dependencies in a different PR, this PR should not do any modifications on the `.opam` file and the `dune-project` file. Is it clear why?

Yes, I understand the reason @pitag-ha !
I'll revert the last commit, and rebase with the master branch, now that the Update-dependencies PR is merged.
Is this correct?

pitag-ha commented:
> I'll revert the last commit, and rebase with the master branch, now that the Update-dependencies PR is merged. Is this correct?

Perfect, thank! Yes, that's correct. Once you've gotten rid of the `.opam` and `dune-project` file modifications, you probably don't even need to rebase yourself given that there shouldn't be any conflicts and so github should be able do a fast-forward workflow. If you want to go the sure way, rebasing on your side sounds good!

AryanGodara commented:
> > I'll revert the last commit, and rebase with the master branch, now that the Update-dependencies PR is merged. Is this correct?
> 
> 
> 
> Perfect, thank! Yes, that's correct. Once you've gotten rid of the `.opam` and `dune-project` file modifications, you probably don't even need to rebase yourself given that there shouldn't be any conflicts and so github should be able do a fast-forward workflow. If you want to go the sure way, rebasing on your side sounds good!

Yes, I reverted last commit and rebased with origin/main.

Is this PR ready now?

pitag-ha commented on cardio_crumble.opam:
```diff
@@ -15,6 +15,8 @@ depends: [
   "portmidi" {>= "0.1"}
   "cmdliner" {>= "1.1.1"}
   "odoc" {with-doc}
+  "portmidi"
+  "cmdliner"
```
These are the two lines that still need to be removed.

AryanGodara replied:
Removed!

pitag-ha reviewed (approved):
Perfect, thanks, @AryanGodara!

Btw, one of these days, I'll try to find time to set up a basic CI checking at least whether the code builds and formats.

AryanGodara commented:
> Perfect, thanks, @AryanGodara!
> 
> 
> 
> Btw, one of these days, I'll try to find time to set up a basic CI checking at least whether the code builds and formats.

Can I try this @pitag-ha ?
I have a bit of experience with github actions, and CI, and can implement something similar to ocaml-mbr repository CI checks
(basic linting and building test at first, then add to it)
